### ISSUE TITLE
Accepting PRs and reviewing code

### ISSUE THREAD

**Issue by bnmnetp** [skulpt/skulpt#414]
I'm moving this Here from #410 to get more people involved in the conversation.

**Comment by bnmnetp** [comment]
I would really prefer to have a system where we had at least two people looking at PRs and concluding that yes, they are good enough to be accepted.  Given that we have two yes votes and no real dissenting votes then any of a few people could accept the PR. 

A lot of projects have a group of core committers.  I think there is a reasonably obvious subset of people who have contributed to skulpt over time that would form a nice group of committers.  This group has some decent variation in background and skills, and has demonstrated to me over time that they know what they are doing.  I will refrain from naming names at the moment and instead ask for volunteers to step forward.  

If other people have experience on other projects where things are run successfully with a different model I would be happy to hear about that too.

**Comment by bnmnetp** [comment]
Just drawing certain people's attention to this issue:  @albertjan @eah13 @waywaaard @rixner @meredydd @bzwheeler @LeszekSwirski

**Comment by albertjan** [comment]
Seems like a sensible direction. :smile: And I would like to *"volunteer"*. Just to make things clear the second person looking at the PR would then accept it if it's up to standards? 

Are we all still in agreement on the contributing [guidelines](https://github.com/skulpt/skulpt/blob/master/CONTRIBUTING.md)? Might be nice to add a section about what a *good* PR looks like.

**Comment by albertjan** [comment]
Interesting read: http://blog.ploeh.dk/2015/01/15/10-tips-for-better-pull-requests/

**Comment by bnmnetp** [comment]
I'm not sure it matters if its the second person that accepts the PR, as long as we have comments from at least two people to the effect that yes, this one is good enough, and nobody is strongly saying, I don't think this should be accepted.

**Comment by bnmnetp** [comment]
Nice article @albertjan.  I experienced some of those things first hand when I made a PR to another repo last week.  I felt like I was running the gauntlet between two guys who could accept or reject the PR.  At times getting conflicting comments from the two of them.

I especially like the "keep it small" approach.  #387 is a great case study.

* It is not small, it is huge 
* it addresses many issues **and** adds a new feature
* In large projects like this I would like to see us create a feature branch, so that we can accept small PR's towards the larger project.
* once all of the work is done, then the feature branch can be merged with master.

Now, the challenge for us in this is that many of these projects arise spontaneously. Someone gets annoyed enough to actually start on the project and so they work on it in private until some progress is made and then create a PR.  I'm not sure how to change that.

I think it might be possible for me to take an existing PR and merge it onto a branch??  I need to research that.

**Comment by eah13** [comment]
I've been pulled away from coding anything but Trinket teaching examples for so long that I'm probably most help continuing cleaning up and tagging issues for now.  Tagging and milestoning requires commit access and I'm very happy to help with those as they're codified.

If it's helpful we could have a weekly or every other weekly review & merge chat where we dedicate time to this and can chat about anything else dev related.  Plenty of devs already work on it weekly and one hour merging pull requests a week would give us sustained momentum when new devs contribute.

Another area I can help: documenting and maintaining the result of discussions like these so it's accessible to the community and expectations are clear.  We can institute a friendly fine for committing straight to master, for instance :)  

In all +1 to the sentiment.  This would be another level of maturity for the project.

**Comment by albertjan** [comment]
I don't think we should try to sync up on merging PR's, I'd say lets just review all the PR's as they come  in and merge them if you're the second reviewer and you both find no issues.

**Comment by meredydd** [comment]
I agree that a synchronous weekly chat would be a bad fit for people for whom Skulpt isn't our "day job" (let alone timezone issues & slowing us down with an explicit merge window).

That said, I definitely support expanding the list of committers/reviewers, and a two-person review makes sense. I am happy to be on that list if you want me, although my time available for Skulpt is sporadic (ie intense blocks irregularly, rather than a certain amount each week). I also agree that it should be made clear that just because new people reviewing PRs does not mean we should push directly to skulpt/master rather than making PRs ourselves - our code needs review too!

My only big concern is that pull requests could fall victim to the bystander effect, in which a tricky PR goes unacknowledged because everyone thinks that someone else will deal with it.

I suggest someone who has a consistent (even if small) amount of time to spend on Skulpt be charged with "directing traffic" - if there's a PR that needs review and nobody's stepped up, pick someone (at random if need be) and tag them, making it Their Responsibility. (Even if they're not the right person to review it, it's now *their* responsibility to tag someone better). @bnmnetp would you be OK taking this role? Some projects have bots to do it, and if it works maybe we could do that too.

**Comment by albertjan** [comment]
@bnmnetp [Here's](https://help.github.com/articles/checking-out-pull-requests-locally/) some documentation about checking out a PR locally

**Comment by bnmnetp** [comment]
Thanks @albertjan  The other option is to simply 

1. Make a branch
2. Close old PR
3. Make a new PR against the branch.

**Comment by bnmnetp** [comment]
@meredydd, as long as I am involved in Skulpt I will continue to push for closing PR's in a reasonably timely manner.  So I am happy to have that role.  The big relief for me will be to have a group of people that I can either tag to take a look at PR's I'm uncertain about, or that people will just go ahead and review new PRs on their own as part of their contribution to the project.

As much as I like the idea of a weekly chance to chat with people involved in the project, living in Malta these last five months has taught me a lot about working across timezones!  I am still hopeful that we can use gitter for asynchronous informal conversations about things that don't necessarily need to be recorded forever as part of an issue or PR.

**Comment by rixner** [comment]
I agree that it would be great to have two pairs of eyes look at a pull request before it gets accepted.  I think the person actually accepting it, though, should still do a little due diligence.  @bnmnetp it would be great if you educated the rest on what you do before accepting a pull request, in terms of running it locally, etc.  I think it would be good if we all at least tried to follow some similar process.

I also like the article @albertjan cites.  However, I will say that I disagree with the notion that the history has to be perfectly clean.  I think there is a lot of emphasis sometimes on cleaning up the history, which goes against one of the advantages of version control.  Cleaning up the history just to make it look better is a big opportunity for introducing bugs.  I don't think you can really review a pull request properly without examining the diff of the entire request, so the history is largely irrelevant.

Even though I can commit directly to master, I have avoided doing so.  Even if I accept my own pull request, I feel that it at least alerts people to the fact that something is being added, and they can take a look at it if they would like.  So, I agree that there is enough activity now that we should probably all avoid committing without a pull request.

One point of process that I think should be clarified, though.  Presumably if a committer makes a pull request, two others still need to review it?  Or are we trusting the committers such that only one other person needs to review it?

I am against weekly chats, because I think it will just be too hard to get everyone to be available at the same time and it will just slow things down because of that...

**Comment by bnmnetp** [comment]
Yes,  I will write up a bit about what I do and add it to the CONTRIBUTING document about what I do before I accept a PR.  It varies of course by what I perceive to be the "riskiness" of the PR, how complex it is, and how well I think I understand it! 

For the sake of discussion here is what I usually do:

1.  I look at the diffs for each file on github, if its pretty obvious what they are doing then great.
2.  I look at the tests provided with the PR and try to think if there are additional tests that would provide better coverage.  Sometimes I might just add them myself or I may send it back to the PR author.
3.  pull the the PR down to my local machine and run all the tests manually locally.  (Unless it is really trivial)
4.  If the code looked a bit more complicated in the diffs, then I will bring it up in my editor and look over the code in context and try to understand it better.
5.  If it is something that I am especially concerned about I will take the built copies of skulpt.min.js and skulpt-stdlib.js and move them into my Runestone project and go through the examples I have on the various pages of the books.  This proved to be very useful in uncovering problems with the new turtle implementation for example.  
6.  With the new style tests, I would like to require the submitter to file bugs for tests that they comment out.

-------

If a committer submits a pull request, I would be OK with said committer making the call on 1 versus 2 other people looking at the PR.  For example:

*  #416 contains just a couple of very simple changes and I would say that one other committer looking at that and then accepting it would be fine.   
* For the new time module PR I would definitely like to have two independent looks, especially since it was a first PR from someone newer to the project.  I took @waywaaard's comment of "looks good" as a sign that he had looked it over.  
* #387 should definitely have two other looks even though I would be happy to  have @waywaaard on the committers list.  But, it adds to the core as well as makes changes to the core.

So I guess I'm mostly saying that 2 should be the rule but lets not be slavish for simple things.

**Comment by LeszekSwirski** [comment]
I agree with what appears to have developed out of this conversation: 2 reviewers per PR, no weekly meetings, asynchronous discussion on gitter, the concept of "assigning" PRs to people. I think it would be helpful if reviewers (especially the first one) could comment on what they checked: "the diff looks fine to me" is very different from "I checked out a copy and ran the tests".

I'm in a similar boat to @meredydd, where I can contribute sporadically but can't make any commitments, but I'd be happy to be considered a part of the team (even if it's going to be an often absent one).

**Comment by albertjan** [comment]
Do we delete the branch after merging the PR?

**Comment by bnmnetp** [comment]
Yes, we do not want to get a bunch of old branches hanging around.  I think either the person who accepted the PR or the originator can delete the branch.  This is only going to be an issue when the PR comes from a branch on skulpt/skulpt rather than someone's Fork.

**Comment by bnmnetp** [comment]
I did create a PR with an update on the committers and my acceptance process.

Anyone want to comment or accept the PR?

**Comment by bnmnetp** [comment]
Here's my bot for the current influx of PR's

```
#!/usr/bin/env python3
import random

print(random.choice(['bnmnetp','rixner',
  'albertjan','bzwheeler',
  'meredydd','leszekswirski']))
```

**Comment by bnmnetp** [comment]
A couple of other things I'm thinking about, now that I'm forced to actually think about it :smile:

* ``xxx.v`` versus ``Sk.ffi.remapToJs(xxx)``   --  it seems like we should adopt a standard, and encourage people to follow it.
* return values on functions that do not return anything.  We ought to look for this and make sure the function is returning ``Sk.builtin.null.null$``
* Be on the lookout for functions that do not remap their parameters or return types.  I think this works in many fewer cases now than it used to, but in some cases where the function is mostly exercised internally but available externally this can slip through the cracks.

## Packages versus Modules

* We ought to clean up our act on this front.  I think I started us down a bad path with a blog post several years ago now.  And now having a much better understanding of packages and modules in real python, I see that we should be following suit here as well.  For example  collections  is a module.  We **should not** artificially make it a package by creating a directory and putting all the code in __init__.js inside said directory.  We **should** simply make file called collections.js in the lib directory.  This already works by the way so it is a simple matter of some renaming.    This change would do wonders for my sanity as I get so confused having 5 different __init__.js files in my tab list.

**Comment by meredydd** [comment]
I agree on remapping and `pyCheckArgs` - although it would be nice to expand that to do type checking/enforcement too (eg "I expect this to be called with one argument, and it's a str").

Possibly this belongs in a separate issue, though?

**Comment by waywaaard** [comment]
Agreed, type checks should be on the list.

**Comment by eah13** [comment]
the packages vs modules cleanup sounds like a great way to go.  I've definitely been confused at times looking thru skulpt source.  seems like open PRs would need to eat a lot of merge conflicts once this gets merged, though, so will need some coordination.  And maybe an addition to `contributing.md`?  Then again since everything's pretty much stubbed out contributors may not need to know since they'll be hacking in a pre-created file.

**Comment by bnmnetp** [comment]
In my experience, git is smart enough to follow the file through a move and rename.  So I'm not sure that would result in the merge conflicts you fear.

**Comment by eah13** [comment]
@bnmnetp a 'Feature Branch' label may be useful for feature branch PRs going forward. Could help designate those PRs as longer running and help reviewers know what they should/shouldnt look to merge

**Comment by bnmnetp** [comment]
Nice idea.  So, I did it.  I guess you could have done it as well.  I think that is very handy flag to let reviewers know that something is a work in progress, and not for immediate review/approval.

**Comment by bnmnetp** [comment]
With our new direction for documentation.  I would like to add a requirement to the review list.  Which is that before we accept a new PR any code touched by the new PR should have an appropriate jsdoc comment if it is a function that we would consider part of the "development API"  

* A function that a web developer might call
* A function that someone working on the core of Skulpt might call
* A function that someone developing a new module might call.

I would also request that if you are working in one of import.js, function.js, abstr.js, or misceval.js that if you take a moment to look for functions there that are not documented but should be.

If we distribute the workload I think we will have a lot of what should be documented covered without doing some massive documentation PR.

**Comment by meredydd** [comment]
Reawakening this thread: Discussion in #835 pointed out that our "two reviews per PR" appears to have slowed our velocity down a _lot_.

I'd like to propose relaxing things at least as far as: "If the change is coming from a committer and it's medium-sized or smaller, one review is enough".

**Comment by bnmnetp** [comment]
The number of actual committers we have today is really quite small.  There are some people like @ebertmi and @LeszekSwirski that haven't committed anything in a couple of years.

I think that if we are going to make good progress toward python 3 we ought to plant a flag in the ground at about where we are right now, at a very good and stable version in my opinion.  We can tag it, I can do a release.  (probably) . And then we should err on the side of gaining velocity towards better python3 coverage.

What do you all think of that?

**Comment by LeszekSwirski** [comment]
Sorry, life got busy :)

**Comment by rixner** [comment]
I am happy to follow whatever procedure we can agree upon.  @bnmnetp planting a flag in the ground now declaring the current version stable seems like a reasonable approach to me.  Or, I suppose, we could start a python3 development branch with a lower bar for acceptance into that branch?

I will make one statement, though: the size of the pull request doesn't necessarily quantify its impact.  For example, some pull requests result in significant slowdowns in Skulpt, despite not being very large.  So, a small pull request can still have a big negative impact from my perspective.  My larger point is this: we all have different objectives in our use of Skulpt.  Having two pairs of eyes has been somewhat effective at balancing our competing objectives and finding small issues with pull requests.

The flip side is that it clearly has slowed down development, especially as the level of activity of different people (including myself) waxes and wanes.  And the issues will likely be found and fixed later.  I can also see the argument that getting to better Python 3 coverage trumps things like performance (at least in the short term).   However, I fear that probably makes it my problem later to try to improve performance if it remains important to me. :smile:

**Comment by bnmnetp** [comment]
Sorry @LeszekSwirski I wasn't meaning to pick on anyone.  Committee work and teaching overloads definitely cut into my own time to contribute as well.

**Comment by bnmnetp** [comment]
@rixner and others.  I know its hard to predict where a slowdown may occur.  But could we build in some kind of performance test into our test suite that would at least alert us to a major performance hit?  I thought of timing the overall test suite, but we are usually adding tests, so that makes it difficult.  But maybe there are a couple of "real world" type benchmarks that we could build in.  Along with a way to store past runs for comparison.

**Comment by rixner** [comment]
@bnmnetp you probably can't keep a record of past performance, as it depends on the machine.  So, you would have to checkout a previous commit and run the performance test there to compare it to the current commit.  This is the process I follow.  Before pulling in upstream changes, I always compare the performance of my current version with the performance of the upstream version.  #827 , for example, led to a 20-30% slowdown on a few different programs I tried.

I think the question is whether performance is a priority for the project.  The answer seems to have generally been "not really".  Measuring it would be nice so that everyone can see the impact of any particular pull request, but would we actually act upon that information?

I know I used to repeat this a lot.  I stopped so as not to annoy everyone.  But, given that we're having this conversation, I would like to make it clear exactly what the situation is from my perspective.

Once upon a time, I made the following gist to talk about performance: https://gist.github.com/rixner/f79dfa52a0385e5c82d147ae3d06c50c

If I run this now, it appears to be around 3x faster on www.codeskulptor.org than on skulpt.org.  This is because I don't deploy most of the new work (because I can't accept such slowdowns).

This massive slowdown seemingly does not bother anyone else, so perhaps it is not worth talking about it, but I do want to make it clear that it's not a small thing.  I need the new work for Python 3, so I now have to think about what such a slowdown means for me.

**Comment by bnmnetp** [comment]
To be clear, skulpt.org is WAY WAY behind.  I haven't released anything to skulpt.org in a long time.  So long that our scripts don't work because Google has changed the whole install process for appengine.

I think it hasn't changed since we had the big discussion about moving to github.io . which didn't happen because I need Scott's help in changing the DNS settings. 😁

**Comment by bnmnetp** [comment]
I just ran it on the latest stuff on master and the time is time elapsed: 13.6239511967 on my 2018 Macbook Pro.

**Comment by rixner** [comment]
That's just depressing.  I ran the current master vs. the skulpt that I have deployed.  The performance difference: 7.4x.

**Comment by rixner** [comment]
I have a few performance-based pull requests that I will try to get to soon, but it's not going to come anywhere near closing that gap.  Regardless, I'm all for moving faster.  I just wish we could do so without continuing to degrade performance.

**Comment by bnmnetp** [comment]
The 13.6 time was me doing `./skulpt run field.py` So, that isn't a fair comparison against the browser.  In the browser the latest runs 2.3 for me.

2.3 versus .9 seconds for your example program does not feel like much of a difference for running a program in the browser.  I guess its because most of what I do in CS1 is not very computationally intense.

How do other projects deal with this?  It seems like an approach of getting the new feature and then making it fast is one path we could follow to ensure we are getting to Python 3 sooner rather than later.

**Comment by rixner** [comment]
I'm seeing 3-7x differences.  But, regardless, even if it's just 2.5x, that's still an enormous difference.  I do two things in my classes where this matters.  First, we build interactive video games.  Even 20-30% slowdowns can mean the difference between a game being fun and not.  A 2.5x slowdown can lead to it being unplayable.  Second, we sometimes do some computationally intense things.  Some students write code that can end up taking 20 minutes to complete.  The difference between 20 minutes and 50 minutes is pretty big.

That program is just one small example and I have used it to gauge how much slowdown I'm likely to see on various other programs.

As I said, though, I guess it's a matter of whether performance is important to the project overall or not.

**Comment by meredydd** [comment]
I am 100% behind the idea of adopting a standard performance test we can apply, and make part of our standard test suite so we generate historical data.

(In fact, once we have a script, it would take only a few lines of Python to apply this to, eg, every PR merge in the last year, and spot which PRs created this slowdown! If we expose this data and make it a one-liner to detect performance issues, we're more likely to nobble them quickly at design/build time.)

**Comment by bnmnetp** [comment]
FYI -- @skulpt/committers I got appengine running on my dev machine and I updated skulpt.org !

**Comment by albertjan** [comment]
@meredydd could the killableWhile and killableFor changes have affected performance even if they're off?

**Comment by albertjan** [comment]
Also on chrome it takes 1.8 seconds and on our build of v8 13 there's something up there too 😄

**Comment by albertjan** [comment]
@rixner your changes I just merged to master don't really have a visible change to performance in chrome but on our v8 build it takes the field runtime from 13~ to 10~ seconds. Pretty good results I'd say.

**Comment by rixner** [comment]
Yes, I've seen similar results in many places. Which, as you say, is pretty good for such a simple change! :)

There are two complexities here.  First, there are so many different run time systems.  Second, you have to get all of the optimization killers before some optimizers will do anything.  So even in cases where that didn't help much, it is still a necessary precursor.

**Comment by LeszekSwirski** [comment]
Note that some of the old "optimization killers" aren't as killer as they used to be since V8 switched to the TurboFan compiler. Might be worth updating the v8 binary, or even removing it entirely and using node or https://github.com/GoogleChromeLabs/jsvu

**Comment by rixner** [comment]
@LeszekSwirski yes, but I still see performance differences, so these things do still matter, even if not as much.   The v8 we're using is pretty clearly no longer representative of browser performance, though...

**Comment by LeszekSwirski** [comment]
Absolutely they matter, often it'll generate worse code or trigger fallbacks into the runtime, I'm just saying it's no longer a complete killer (i.e. won't disable optimisation entirely), and you should thus be able to see an effect from incremental changes.
